wire-ios-data-model/Source/ManagedObjectContext/Migration/118-119/FixDuplicateOneOnOneConversationsAction.swift

```diff
@@ -0,0 +1,128 @@
+
+        // 1) fetch all conversations to fix
+        guard let fake1On1Conversations: [ZMConversation] = try? fakeOneOnOneConversations(context: context), !fake1On1Conversations.isEmpty else {
+            return
+        }
+
+        let fakeOneOnOnePerUser = conversationsPerUser(from: fake1On1Conversations)
+
+        for (user, conversations) in fakeOneOnOnePerUser {
+            guard conversations.count > 1 else {
+                // if only one conversation, nothing to fix
+                continue
+            }
+            var sortedConversations = conversations.sorted { lhs, rhs in
+                lhs.primaryKey < rhs.primaryKey
+            }
+
+            // 2) set 1:1 conversation
+            let theOneOnOneConversation = sortedConversations.removeFirst()
+            user.oneOnOneConversation = theOneOnOneConversation
+
+            WireLogger.localStorage.debug("Fixing oneOnOne conversation", attributes: [.conversationId: theOneOnOneConversation.remoteIdentifier.safeForLoggingDescription])
+
+            // 3) select all messages from other conversations
+            for otherConversation in sortedConversations {
+                moveMessages(from: otherConversation, to: theOneOnOneConversation)
+            }
+        }
+    }
+
+    private func conversationsPerUser(from conversations: [ZMConversation]) -> [ZMUser: [ZMConversation]] {
+        var fakeOnOnOnePerUser = [ZMUser: [ZMConversation]]()
+        for conversation in conversations {
+            if let otherUser = conversation.localParticipantsExcludingSelf.first {
+                if fakeOnOnOnePerUser[otherUser] == nil {
+                    fakeOnOnOnePerUser[otherUser] = [conversation]
+                } else {
+                    fakeOnOnOnePerUser[otherUser]?.append(conversation)
+                }
+
+            }
+        }
+        return fakeOnOnOnePerUser
+    }
+
+    private func moveMessages(from otherConversation: ZMConversation,
+                              to theOneOnOneConversation: ZMConversation) {
+        otherConversation.allMessages.forEach { message in
+
+            // 4) set them to conversation 1:1
+            if message.visibleInConversation != nil {
+                message.visibleInConversation = theOneOnOneConversation
+            }
+            if message.hiddenInConversation != nil {
+                message.hiddenInConversation = theOneOnOneConversation
+            }
+        }
+
+        // get the draft message if available and none present
+        if theOneOnOneConversation.draftMessageData == nil {
+            theOneOnOneConversation.draftMessageData = otherConversation.draftMessageData
+        }
+        if theOneOnOneConversation.draftMessageNonce == nil {
+            theOneOnOneConversation.draftMessageNonce = otherConversation.draftMessageNonce
+        }
+    }
+
+    private func fakeOneOnOneConversations(context: NSManagedObjectContext) throws -> [ZMConversation] {
```

Comment:
Does it make sense to define once and reuse it?

Reply:
not sure to follow, it breaks down the length of initial function

Reply:
I think I was referring to if we need to or should factor out the predicates that define a fake one on one, so we don't have multiple implementations of it (I can't remember now if we do or not.

Reply:
we kind of do inside the [migrationAction](https://github.com/wireapp/wire-ios/blob/4e7f7d2a795a49646dacf82b8c8257a729eeefde/wire-ios-data-model/Source/ManagedObjectContext/Migration/113-114/OneOnOneConversationMigrationAction.swift#L85)  but it slightly different.

Here I select all OneOnOneConversations and on the OneOnOneConversationMigrationAction we select all oneOnOneConversations with a given user.

For now I would keep this similar implementation as I don't see benefits or how to factor this out